Update Hexagon asm and analysis plugins

Rot127 opened the pull request:
Replaces https://github.com/rizinorg/rizin/pull/1338

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Updates Hexagon `analysis` and `asm` plugin.

**Supports**
- Add `asm` support for v5-v68 and HVX instructions
  - Recognize instruction packets.
- Basic analysis support
  - set instruction type (return, call, jumps hardware loop etc.)
  - Enables search for immediate operands.
- Patching of relocs 

**Not yet implemented**
- ~~System registers~~
- ~~[System/Monitor instructions](https://github.com/rizinorg/rz-hexagon/issues/21)~~
- [Can not analyse >1024bit vector registers](https://github.com/rizinorg/rizin/issues/1287)
- No ESIL support
- [~~No tests yet for HVX instructions and hardware loops.~~](https://github.com/rizinorg/rz-hexagon/issues/8)
- Test files for reloc patching

**Test plan**

See test files.

**Closing issues**

closes https://github.com/rizinorg/rz-hexagon/issues/13
closes https://github.com/rizinorg/rz-hexagon/issues/15
closes https://github.com/rizinorg/rz-hexagon/issues/17
closes https://github.com/rizinorg/rz-hexagon/issues/18
closes https://github.com/rizinorg/rz-hexagon/issues/19
closes https://github.com/rizinorg/rz-hexagon/issues/21
closes https://github.com/rizinorg/rz-hexagon/issues/22
closes https://github.com/rizinorg/rz-hexagon/issues/24

Rot127 committed:
Add support for: v5-v68+HVX, System instructions/register; Change plugin design to monolithic; Basic ELF-reloc patching.

- Update instruction set to v5-v67 + HVX.
- Introduce monolithic plugin architecture to reverse opcodes in the same functions.
  - Analysis and Asm functionality was previously seperated and let to inconsistensies and incorrectly reversed opcodes.
- Add missing system instructions and registers.
- Enable search for immediates.
- Mark packets of instructions.
- Set more analysis data in RzAnalysisOp.
- Add basic ELF-reloc patching.
- Fix several disassembly and analysis bugs.
  - Jumps were not aligned to packet begin.
  - Endloop packets were not marked as such.
  - Rs.New registers were incorrectly disassembled.
  - Calling convention was out of date.
- Add several formatting configs for the produced disassembly.
  - Dis/Enable UTF-8.
  - Print or omit # prefix of immediates.
  - Print signed immediates with signe or as unsigned integers.

Rot127 commented:
I have absolutely no idea what the cause of the failing test is. I just created a VM and built rizin as the ubuntu-tcc-test Action does it. Basically copied and pasted the commands to build TinyCC and meson.

All tests succeeded. No error.

The only difference is, that I use a debian-10 in the VM. No Ubuntu. Unfortunately I can't install a Ubuntu VM at the moment and my enthusiasm debugging it is meanwhile really low also.

Do you have any advice how to proceed?

Just in case I'm a blockhead. The commands I compiled rizin with TinyCC:

```
git clone https://repo.or.cz/tinycc.git
cd tinycc
#git checkout mob
# Well-tested commit
git checkout 675046bd59bc6977bb2016c7d2e115ace8a6ae6c
chmod +x ./configure
./configure --prefix=/usr
make
sudo make install
sudo apt-get --assume-yes install python3-wheel python3-setuptools python3-pip
sudo python3 -m pip install ninja
git clone https://github.com/ret2libc/meson.git
cd meson
git checkout tiny-cc
python3 setup.py build
sudo python3 setup.py install
cd ~
git clone https://github.com/Rot127/rizin.git
cd ~/rizin/
git checkout Hexagon
env CC=tcc meson --prefix=/usr build
ninja -C build && sudo ninja -C build install
mkdir -p test/bins/elf/analysis/
cd test/bins/elf/analysis/
wget https://raw.githubusercontent.com/rizinorg/rizin-testbins/master/elf/analysis/hexagon-hello-loop
cd ~/rizin
rz-test test/db/analysis/hexagon 

Running from /home/user/rizin/test
Loaded 3 tests.
Skipping json tests because jq is not available.
[3/3]                       3 OK         0 BR        0 XX        0 FX
Finished in 0 seconds.
```

Rot127 commented:
So I ran the tests on an Ubuntu 20.04.3 LTS VM and they succeeded. It must be something with Github Actions than.

wargio commented:
> So I ran the tests on an Ubuntu 20.04.3 LTS VM and they succeeded. It must be something with Github Actions than.

also with tinycc?

Rot127 commented:
Oh, I was stupid. Sorry.
Ran `meson --prefix=/usr build` instead of `env CC=tcc meson --prefix=/usr build`.
Now the tests fail.

Will debug it once I have more time.

ret2libc commented on librz/bin/p/bin_elf.inc:
```diff
@@ -387,13 +387,323 @@ static ut64 reloc_targets_vfile_size(RzBinFile *bf, ELFOBJ *obj) {
+			//	val = (G);
+			//	break;
+			//case R_HEX_GOT_32_6_X:
+			//	word = 4;
+			//	val = (G) >> 6;
+			//	break;
+			//case R_HEX_GD_GOT_32_6_X:
+			//	word = 4;
+			//	val = (G) >> 6;
+			//	break;
+			//case R_HEX_IE_32_6_X:
+			//	word = 4;
+			//	val = (G + GOT) >> 6;
+			//	break;
+			//case R_HEX_IE_GOT_32_6_X:
+			//	word = 4;
+			//	val = (G) >> 6;
+			//	break;
+			//case R_HEX_LD_GOT_LO16:
+			//	bitmask = 0x00c03fff;
+			//	word = 4;
+			//	val = (G);
+			//	break;
+			//case R_HEX_LD_GOT_HI16:
+			//	bitmask = 0x00c03fff;
+			//	word = 4;
+			//	val =  (G) >> 16;
+			//	break;
+			//case R_HEX_LD_GOT_32:
+			//	word = 4;
+			//	val = (G);
+			//	break;
+			//case R_HEX_LD_GOT_16:
+			//	word = 4;
+			//	val = (G);
+			//	break;
+			//case R_HEX_LD_GOT_32_6_X:
+			//	word = 4;
+			//	val = (G) >> 6;
+			//	break;
```
remove if not needed.

ret2libc reviewed (changes requested):
I haven't really followed here... but if @wargio things is good, I'm ok to merge this as it is anyway quite confined to hexagon code. Please address just this thing first.

XVilka commented on librz/analysis/p/analysis_hexagon.c:
```diff
@@ -1,6 +1,12 @@
-// SPDX-FileCopyrightText: 2018 xvilka <anton.kochkov@gmail.com>
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+//
```
Remove this empty line please.

XVilka commented on librz/analysis/arch/hexagon/hexagon_analysis.h:
```diff
@@ -1,4 +1,10 @@
-// SPDX-FileCopyrightText: 2018-2021 Anton Kochkov <anton.kochkov@gmail.com>
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+//
```
Remove this empty line please

XVilka commented on librz/asm/arch/hexagon/hexagon.c:
```diff
@@ -1,237 +1,987 @@
-// SPDX-FileCopyrightText: 2018-2021 Anton Kochkov <anton.kochkov@gmail.com>
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+//
```
Remove this empty line please

XVilka commented on librz/asm/arch/hexagon/hexagon.c:
```diff
@@ -1,237 +1,987 @@
+}
+
+/**
+ * @brief Applies the last constant extender to the immediate value of the given HexOp.
```
Please use `\brief` syntax everywhere: https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#documentation

XVilka commented on librz/asm/arch/hexagon/hexagon.h:
```diff
@@ -1,9 +1,16 @@
-// SPDX-FileCopyrightText: 2018-2021 Anton Kochkov <anton.kochkov@gmail.com>
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+//
```
Remove this empty line please

XVilka commented on librz/asm/arch/hexagon/hexagon.h:
```diff
@@ -46,237 +51,521 @@ typedef enum {
 typedef struct {
 	int instruction;
 	ut32 mask;
-	HexPred predicate; // predicate set if set
```
Why are predicates no longer in use? They are pretty common in the code.

Rot127 replied:
Oh, I quite frankly forgot about them. Will implement as soon as possible.

Rot127 replied:
Fixed with https://github.com/rizinorg/rizin/pull/1614/commits/1435be081c4e3265aa820706993112eb78670499

Rot127 replied:
Although https://github.com/rizinorg/rz-hexagon/issues/29 still persists.

XVilka commented on librz/bin/p/bin_elf.inc:
```diff
@@ -387,13 +387,323 @@ static ut64 reloc_targets_vfile_size(RzBinFile *bf, ELFOBJ *obj) {
 	ut8 buf[8];
 	switch (e_machine) {
+	case EM_QDSP6:;
+		ut8 word = 0;
```
Can you please extract the reloc patching for Hexagon in a separate function? We are working on improving and refactoring relocations patching in ELF and it will make that easier in the future:
- https://github.com/rizinorg/rizin/pull/1699
- https://github.com/rizinorg/rizin/issues/1141

XVilka commented on librz/bin/p/bin_elf.inc:
```diff
@@ -387,13 +387,323 @@ static ut64 reloc_targets_vfile_size(RzBinFile *bf, ELFOBJ *obj) {
+		ut64 bitmask = UT64_MAX; // Mask for bits in the opcode which get patched.
+		switch (rel->type) {
+		default:
+			eprintf("Patching for reloc type %d not implemented.", rel->type);
```
Use `RZ_LOG_ERROR()`

XVilka reviewed (changes requested):
Overall looks good. What bothers me - why you don't use predicates? They are pretty important.
Also, for relocations - could you please add a test specifically for relocations? Like it's done in https://github.com/rizinorg/rizin/pull/1699/files#diff-815fd301b119045a3c4aef6ecbb05174a1d997c1c43cf2b65d4a461ea57bd7b4R114
So for every different type of relocation you implemented, if it's in the test binary - print bytes, print instruction. It's recommended to cross-check your results against `readelf`, `objdump`, IDA Pro, Binary Ninja, Ghidra.

Good job!

Rot127 commented:
@XVilka 

> Good job!

Thank you. I really appreciate the work you guys put into reviewing this.

> What bothers me - why you don't use predicates? They are pretty important.

I quite frankly forgot about them. Will start with the implementation today.

Can you give me a hint where rizin uses this information? I can't find a member in `RzAnalysisOp` which represents predicates properly. But maybe my understanding is limited.

> It's recommended to cross-check your results against readelf, objdump, IDA Pro, Binary Ninja, Ghidra.

| Tool | comment |
|-------|--------------|
| objdump |  I checked the disassembly and analysis test against the SDKs `objdump`. And actively advice it in the test documentation in the rz-hexagon docs. 
| IDA Pro | I can't effort a license for IDA Pro, therefore not test against it. That's the reason why I started developing the Hexagon plugin for rizin in the first place. |
| Binary Ninja | Same like IDA Pro. I can't effort a license and the demos can't load the [Hexagon plugin](https://github.com/google/binja-hexagon). |
| Ghidra | [Ghidra does not support the Hexagon architecture yet](https://github.com/NationalSecurityAgency/ghidra/issues/1595). 

The problem for me is, that tests implementations need a lot of time which I currently not have (e.g. compiling HVX code with the SDK is buggy, building a ELF binary with all implemented relocations is quite hard, as I just start learning about relocations details etc.).
At the moment I can, unfortunately, only promise that I will add them gradually once I use it more myself. For me it is absolutely fine if this PR is not merged yet for this reason. The plugin will be under development at least for the next 5-6 months anyways.

Rot127 commented:
Almost forgot. Testing the reloc patching is quite hard at the moment since the rebasing is so broken.

XVilka commented:
@Rot127 do you mean this one? https://github.com/rizinorg/rizin/issues/1661
cc @ret2libc

Rot127 commented:
> @Rot127 do you mean this one? #1661

Yes. It does correctly patch if the binary is loaded with `rizin -m <baddr>`. But because of #1661 and building a binary which uses all implemented relocations is quite time consuming I postponed the tests.

XVilka commented on librz/bin/p/bin_elf.inc:
```diff
@@ -387,13 +387,235 @@ static ut64 reloc_targets_vfile_size(RzBinFile *bf, ELFOBJ *obj) {
 }
 
-static void patch_reloc(struct Elf_(rz_bin_elf_obj_t) * obj, RzBinElfReloc *rel, ut64 S, ut64 B, ut64 L) {
+static void patch_reloc_hexagon(RZ_INOUT RzBuffer *buf_patched, const int rel_type, const ut64 patch_addr, const ut64 S, const ut64 B, const ut64 L, const ut64 GOT, const ut64 A, const ut64 P) {
```
Instead of passing so many parameters I suggest just create some context structure to group them into one or a few structures and pass them instead.

XVilka commented on librz/include/rz_bin.h:
```diff
@@ -744,6 +744,20 @@ struct rz_bin_reloc_storage_t {
 	size_t target_relocs_count;
 }; // RzBinRelocStorage
 
+typedef struct rz_bin_reloc_formular_symbols_t {
```
I don't think it should be in the `rz_bin.h`. Could you please move it back to the `bin_elf.inc`? It's ELF-specific, not needed for anything else.

XVilka commented:
Please also rebase on top of the latest `dev` to resolve the conflicts.

Rot127 commented:
`TinyCC build` test fails because of https://github.com/rizinorg/rz-hexagon/issues/24.
Besides, `tcc-builds` fail on real world binaries. The static variables get always incorrectly set. But I can't find the issue.

It seems to me that this is a pretty annoying `tcc` bug. Do you have any advice how to proceed with this? Guess it can't be merged if it does not work with `tcc`?

XVilka commented:
@Rot127 usually TinyCC devs fix the error when reported provided you can minimize the reproducer. See their mailing list:
- https://lists.nongnu.org/mailman/listinfo/tinycc-devel
- https://lists.nongnu.org/archive/html/tinycc-devel/

Then, once they fixed we can bump the used commit in [.github/workflows/tcc.yml:49](https://github.com/rizinorg/rizin/blob/dev/.github/workflows/tcc.yml#L49).

Rot127 commented:
https://github.com/rizinorg/rizin/pull/1614/commits/52d6deb1645f3682dd941c31ac6e4180c2862aed changes the architecture of the plugin pretty substantially.
It closes https://github.com/rizinorg/rz-hexagon/issues/24 and is hopefully useful for the future (if RzArch is introduced). 

An opcode is now reversed by only one function (in hexagon_disas.c).
The flow graph would look a bit like that:

```
                                   <--------- writes RzAsmOp ---------------------------------------+
rz_do_disas()-----+                                                                                  |
                  |----> reverse_opcode() -----> |add empty insn to global state| -----> do_disas_and_analysis(insn)
rz_do_analysis()--+                                                                                  |
                                  <------- writes RzAnalysisOp--------------------------------------+
```

Also added in this commit:
- Fix: `Rs.new` registers get recognized correctly. 
- Fix: Jump offsets are now added to the packet address, not the address of the instruction itself (which was incorrect).
- Simple jumps are now marked as block end.

XVilka commented:
@wargio @ret2libc @thestr4ng3r please take a look in regard of keeping the same state between analysis and disassembly plugins.

wargio commented:
Damn.. we really need RzArch

Rot127 commented:
Is it a problem that the `hexagon_disas.c` is so large that gcc needs >8GB of RAM to build rizin?
My VM fails to build it, if it has 8GB assigned. Need to add a swap file with additional 4GB.

Building with tcc obviously works fine. Besides the issues with the tests.

XVilka commented:
Yes, this is a problem and also a bug in GCC. I recommend reporting it to https://gcc.gnu.org/bugzilla/
Meanwhile, the solution would be to split this file. BTW, how big it is?

Rot127 commented:
> BTW, how big it is?

It is 4.4M and will get 30% to 50% bigger if we add more info to each instruction in the future (RZIL, memory access info etc.). Splitting it now should make each file ~0.14M big

> Meanwhile, the solution would be to split this file.

This pull request gets bigger every day. Three months ago I just wanted to add the new v68 instructions :smiley:  
Anyways. Will split them.

thestr4ng3r commented:
> It is 4.4M

That is quite large, even for generated code. Do you think it would be possible to optimize the code generator, for example splitting the individual instruction in different categories that will then use the same code for certein instruction properties and can thus avoid code duplication.

Rot127 commented:
> for example splitting the individual instruction in different categories

Not without reworking it. The meaning of each bit at position i differs from opcode to opcode (sometimes pretty significantly). There is certainly a logic to it, but it isn't publicly documented. So it probably would need some reversing and al lot more testing.

That's why the generator simply iterates over all instruction defintions of the LLVM and sets the struct members accordingly.

But: [this](https://github.com/n-o-o-n/idp_hexagon) IDA plugin implements the logic. But it very much looks like the person who develops it has access too proprietary Qualcomm documentation.

thestr4ng3r commented:
Hmm I see. What about, instead of generating code for each property, generating a lookup table in C for some properties? That would shift a bit of the work to the runtime, but probably also compress the entire thing a bit.
By the way, what is the code size after the file has been compiled?

Rot127 commented:
> By the way, what is the code size after the file has been compiled?

`ls -l` for `./build/librz/analysis/librz_analysis.a.p/.._asm_arch_hexagon_hexagon_disas.c.o` gives me:
| compiler | target | size |
|---|----|----| 
| tcc | x86_64 |  5.3M |
| gcc | x86_64 | 1.9M |
| clang | aarch64 | 1.4M |

Rot127 commented:
> What about, instead of generating code for each property, generating a lookup table in C for some properties?

Mh, I will ponder about this a bit. But for now I will remove unnecessary function calls from it (~3 per instruction). This should drive the memory usage of gcc down quite a bit.

@thestr4ng3r Just in case you have time and want to take a quick look at it yourself, in `librz/asm/arch/hexagon/hexagon_disas.c @ line 56088` is an instruction which is pretty complex and has a lot of assignments. Maybe you see something right away.

Rot127 commented:
It's weird. But I no longer get the problem with >4gb memory usage of gcc (top shows something around 16MB).
So if you like, you can review it.

XVilka commented:
There are a lot of unrelated changes. Thus, I propose:
1. Squash all commits into one - this way rebasing should be easier
2. Rebase on top of the latest `dev`
3. Fix `clang-format` warning:
```
Ubuntu clang-format version 13.0.1-++20211110063012+9dc7d6d5e326-1~exp1~20211110183538.22
librz/asm/arch/hexagon/hexagon_disas.c:18:26: error: code should be clang-formatted [-Wclang-format-violations]
#include "hexagon_arch.h"
                         ^
```

XVilka reviewed (approved):
Looks good to merge. Please use the better commit name when merging.

XVilka commented:
Merge if https://github.com/rizinorg/rizin/pull/2033 is green (excluding the known debug test ASAN failure)

XVilka commented:
Fails to build on Windows AppVeyor:
```
FAILED: librz/asm/librz_asm.a.p/arch_hexagon_hexagon_arch.c.obj 
"cl" "-Ilibrz/asm\librz_asm.a.p" "-I." "-I.." "-I..\librz\include" "-Ishlr" "-I..\shlr" "-I..\librz\asm\arch\include" "-I..\librz\asm\arch" "-I..\librz\asm\arch\h8300" "-I..\librz\asm\arch\hexagon" "-I..\librz\asm\arch\msp430" "-I..\librz\asm\arch\rsp" "-I..\librz\asm\arch\mcore" "-I..\librz\asm\arch\v850" "-I..\librz\asm\arch\propeller" "-I..\librz\asm\arch\ebc" "-I..\librz\asm\arch\cr16" "-I..\librz\asm\arch\8051" "-I..\librz\asm\arch\v810" "-I..\librz\asm\arch\or1k" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-I..\librz\bin\format" "-I..\subprojects\capstone-bundled\include" "-I..\subprojects\capstone-bundled\include\capstone" "-I..\shlr\spp" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz/asm\librz_asm.a.p\arch_hexagon_hexagon_arch.c.pdb" /Folibrz/asm/librz_asm.a.p/arch_hexagon_hexagon_arch.c.obj "/c" ../librz/asm/arch/hexagon/hexagon_arch.c
../librz/asm/arch/hexagon/hexagon_arch.c(613): error C2375: 'hex_extend_op': redefinition; different linkage
c:\projects\rizin\librz\asm\arch\hexagon\hexagon.h(550): note: see declaration of 'hex_extend_op'
[1116/1727] Compiling C object librz/asm/librz_asm.a.p/arch_h8300_h8300_disas.c.obj
[1117/1727] Compiling C object librz/asm/librz_asm.a.p/arch_hexagon_hexagon_disas.c.obj
ninja: build stopped: subcommand failed.
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/41680793/job/6m3ufxlgdbrqrbi0

XVilka commented:
Failing test with ASAN:
```


[XX] db/analysis/hexagon hexagon jumps
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'b 0x100000
e analysis.jmp.cref=true
s sym.main
aaa
afx
' bins/elf/analysis/hexagon-hello-loop
-- stdout
--- expected
+++ actual
@@ -5,3 +5,7 @@
 C 0x0000513c -> 0x000050f8 [   call sym.pWorld
 c 0x00005140 -> 0x00005144 [   jump 0x5144
 c 0x00005150 -> 0x00005128 ?   jump 0x5128
+WARNING: rz_reg_set_value: assertion 'reg && item' failed (line 145)
+WARNING: rz_reg_set_value: assertion 'reg && item' failed (line 145)
+WARNING: rz_reg_set_value: assertion 'reg && item' failed (line 145)
+WARNING: rz_reg_set_value: assertion 'reg && item' failed (line 145)

-- stderr
Cannot determine entrypoint, using 0x00005000.
[ ] Analyze all flags starting with sym. and entry0 (aa)
```

Rot127 committed:
Add missing RZ_API in declaration.

Rot127 commented:
https://github.com/rizinorg/rizin/pull/1614/commits/6942f6c9083f5f2347f697b95d93ae09f10b0a88 should fix the Windows build.

XVilka commented:
The other PR was green.
